lib/auth/mfa/mfav1/service_test.go

```diff
@@ -0,0 +1,551 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package mfav1_test
+
+import (
+	"testing"
+
+	"github.com/gravitational/trace"
+	"github.com/stretchr/testify/require"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/constants"
+	mfav1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/mfa/v1"
+	"github.com/gravitational/teleport/api/types"
+	apievents "github.com/gravitational/teleport/api/types/events"
+	"github.com/gravitational/teleport/lib/auth/authtest"
+	mfav1impl "github.com/gravitational/teleport/lib/auth/mfa/mfav1"
+	"github.com/gravitational/teleport/lib/authz"
+	"github.com/gravitational/teleport/lib/events"
+	"github.com/gravitational/teleport/lib/events/eventstest"
+	"github.com/gravitational/teleport/lib/tlsca"
+)
+
+// TestCreateSessionChallenge_WebAuthn tests the CreateSessionChallenge method for WebAuthn challenges.
+func TestCreateSessionChallenge_WebAuthn(t *testing.T) {
+	emitter := &eventstest.MockRecorderEmitter{}
+
+	authServer, err := NewMockAuthServer(authtest.ServerConfig{
+		Auth: authtest.AuthServerConfig{
+			AuditLog:    &eventstest.MockAuditLog{Emitter: emitter},
+			ClusterName: "test-cluster",
+			Dir:         t.TempDir(),
+			AuthPreferenceSpec: &types.AuthPreferenceSpecV2{
+				SecondFactors: []types.SecondFactorType{
+					types.SecondFactorType_SECOND_FACTOR_TYPE_WEBAUTHN,
+				},
+				Webauthn: &types.Webauthn{
+					RPID: "localhost",
+				},
+			},
+		},
+	},
+		nil,
+	)
+	require.NoError(t, err)
+
+	service, err := mfav1impl.NewService(mfav1impl.ServiceConfig{
+		AuthServer: authServer,
+		Cache:      authServer.Auth().Cache,
+		Emitter:    authServer.Auth(),
+		Identity:   authServer.Auth().Identity,
+	})
+	require.NoError(t, err)
+
+	user, err := authtest.CreateUser(t.Context(), authServer.Auth(), "test-user")
+	require.NoError(t, err)
+
+	ctx := authz.ContextWithUser(t.Context(), authz.LocalUser{
+		Identity: tlsca.Identity{
+			Username: user.GetName(),
+		}},
+	)
+
+	_, err = authtest.RegisterTestDevice(
+		ctx,
+		authServer.Auth(),
+		"webauthn-device",
+		proto.DeviceType_DEVICE_TYPE_WEBAUTHN,
+		nil,
+	)
+	require.NoError(t, err)
+
+	resp, err := service.CreateSessionChallenge(
+		ctx,
+		&mfav1.CreateSessionChallengeRequest{
+			Payload: &mfav1.SessionIdentifyingPayload{
+				Payload: &mfav1.SessionIdentifyingPayload_SshSessionId{
+					SshSessionId: []byte("test-session-id"),
+				},
+			},
+		},
+	)
+	require.NoError(t, err)
+
+	// Verify a WebAuthn challenge was returned.
+	require.NotNil(t, resp)
+	require.NotNil(t, resp.MfaChallenge)
+	require.NotNil(t, resp.MfaChallenge.WebauthnChallenge)
+	require.NotEmpty(t, resp.MfaChallenge.WebauthnChallenge.PublicKey.Challenge)
```

Comment:
Let's go all the way and validate the Webauthn challenge, instead of picking parts of it to check for nilness. That is the best way to guarantee it's good.

As for the asserts, this suffices if you take the advice above:

```suggestion
	require.NotNil(t, resp.GetMfaChallenge().GetWebauthnChallenge(), "WebauthChallenge must not be nil")
```

Reply:
I still see both Create and Validate test, but I think at this point the Create test isn't doing anything Validate already doesn't cover. (Maybe the CreateSSO is, but the Webauthn one isn't.)